### PR TITLE
JSONPath template language

### PR DESCRIPTION
**Pull request by daizuozhuo**
@brendandburns Thanks. This is the rebased version of #9296 
- [x] "" quote string
- [x] . key operator 
- [x] [] array operator
- [x] start : end : step slice operator
- [x] @ current object
- [x] \* wildcard operator
- [x] ?() array filter
- [x] .. parent operator
- [x] [,] array indices as a set
- [x] range, end keyword

### CODE REVIEW

File: pkg/util/jsonpath/jsonpath_test.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+		{"range addresss", "{.items[*].status.addresses[*].address}", nodesData,
+			`127.0.0.1 127.0.0.2 127.0.0.3`},
+		{"two range", "{range .items[*]}{range .status.addresses[*]}{.address}, {end}{end}", nodesData,
+			`127.0.0.1, 127.0.0.2, 127.0.0.3, `},
+		{"item name", "{.items[*].metadata.name}", nodesData, `127.0.0.1 127.0.0.2`},
+		{"union nodes capacity", "{.items[*]['metadata.name', 'status.capacity']}", nodesData,
+			`127.0.0.1 127.0.0.2 {cpu: 4} {cpu: 8}`},
+		{"range nodes capacity", "{range .items[*]}[{.metadata.name}, {.status.capacity}] {end}", nodesData,
+			`[127.0.0.1, {cpu: 4}] [127.0.0.2, {cpu: 8}] `},
+		{"user passowrd", `{.users[?(@.name=="e2e")].user.password}`, nodesData, "secret"},
```
Comment (bgrant0607) [inline pkg/util/jsonpath/jsonpath_test.go]: Typo?

Comment (daizuozhuo) [inline pkg/util/jsonpath/jsonpath_test.go]: Oh sorry

**Comment by brendandburns** [comment]
LGTM.  Please squash the commits. and then I will add 'lgtm' label and get this merged.

**Comment by daizuozhuo** [comment]
@brendandburns Thanks. I squashed them into 8 commits.

**Comment by brendandburns** [comment]
Sorry, I wasn't clear.  Please squash to 1 commit.  If there really are multiple conceptual things, they should be broken out as separate PRs (but I think it's fine for there to be one commit for this)

Shippable appears to be a flake, I'm re-running.

@k8s-bot test this please

--brendan

**Comment by daizuozhuo** [comment]
@brendandburns Thanks. There is only one commit now.

**Comment by brendandburns** [comment]
LGTM, shippable is a flake, re-running.

**Comment by brendandburns** [comment]
There is something severely borked in shippable.  Its complaining about v1beta3, but there's nothing about v1beta3 in this PR.

@k8s-bot test this please

If this passes jenkins, I'm just going to merge it.

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Comment by daizuozhuo** [comment]
This is strange. The code of 8dcd077 and 949704a are the same, but GCE e2e build/test passed for commit 8dcd077 and failed on 949704a.

**Comment by brendandburns** [comment]
@daizuozhuo I think it's a flake, I'm just going to merge this, and we'll see what happens...

**Comment by brendandburns** [comment]
@daizuozhuo now that this is submitted, are you working on the integration with kubectl?  We should create a new output format printer.

Thanks!
--brendan

**Comment by daizuozhuo** [comment]
@brendanburns Yes, I will send pull request soon.

**Comment by JeanMertz** [comment]
Awesome work here guys, looking forward to this. Given this work, I'm going to stick to Go templating for now, instead of searching for some 3rd-party json pre-processor to fill in variables and the likes.

**Pull request merged**